homeassistant/components/weather/smhi.py

```diff
@@ -0,0 +1,244 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+# Used to map condition from API results
+CONDITION_CLASSES = {
+    'cloudy': [5, 6],
+    'fog': [7],
+    'hail': [],
+    'lightning': [21],
+    'lightning-rainy': [11],
+    'partlycloudy': [3, 4],
+    'pouring': [10, 20],
+    'rainy': [8, 9, 18, 19],
+    'snowy': [15, 16, 17, 25, 26, 27],
+    'snowy-rainy': [12, 13, 14, 22, 23, 24],
+    'sunny': [1, 2],
+    'windy': [],
+    'windy-variant': [],
+    'exceptional': [],
+}
+
+
+# 5 minutes between retrying connect to API again
+RETRY_TIMEOUT = 5*60
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=31)
+
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Old way of setting up components.
+
+    Can only be called when a user accidentally mentions smhi in their
+    config. But even in that case it would have been ignored.
+    """
+    pass
+
+
+async def async_setup_entry(hass: HomeAssistant,
+                            config_entry: ConfigEntry,
+                            config_entries) -> bool:
+    """Add a weather entity from map location."""
+    location = config_entry.data
+    name = location[CONF_NAME]
+
+    session = aiohttp_client.async_get_clientsession(hass)
+
+    entity = SmhiWeather(name, location[CONF_LATITUDE],
+                         location[CONF_LONGITUDE],
+                         session=session)
+    entity.entity_id = ENTITY_ID_SENSOR_FORMAT.format(name)
+
+    config_entries([entity], True)
+    return True
+
+
+class SmhiWeather(WeatherEntity):
+    """Representation of a weather entity."""
+
+    def __init__(self, name: str, latitude: str,
+                 longitude: str,
+                 session: aiohttp.ClientSession = None) -> None:
+        """Initialize the Smhi weather entity."""
+        from smhi import Smhi
+
+        self._name = name
+        self._latitude = latitude
+        self._longitude = longitude
+        self._forecasts = None
+        self._fail_count = 0
+        self._smhi_api = Smhi(self._longitude, self._latitude,
+                              session=session)
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    async def async_update(self) -> None:
+        """Refresh the forecast data from smhi weather API."""
+        from smhi.smhi_lib import SmhiForecastException
+
+        def fail():
+            self._fail_count += 1
+            if self._fail_count < 3:
+                self.hass.helpers.event.async_call_later(
+                    RETRY_TIMEOUT, self.retry_update())
+
+        try:
+            with async_timeout.timeout(10, loop=self.hass.loop):
+                self._forecasts = await self.get_weather_forecast()
+                self._fail_count = 0
+
+        except asyncio.TimeoutError:
+            _LOGGER.error("Failed to connect to SMHI API, "
+                          "timeout, retry in 5 minutes")
+            fail()
+        except SmhiForecastException:
```

Comment:
You can combine this exception with the other timeout exception

Reply:
It was a minor difference, where the first suggested the error being a timeout. Not sure it will be beneficial so I agree. I took your feedback.